benchmark/main.cpp

```diff
@@ -56,979 +50,1198 @@
-    if( todo->aes_ccm )
-    {
-        int keysize;
-        mbedtls_ccm_context ccm;
-
-        mbedtls_ccm_init( &ccm );
-        for( keysize = 128; keysize <= 256; keysize += 64 )
-        {
-            mbedtls_snprintf( title, sizeof( title ), "AES-CCM-%d", keysize );
-
-            memset( buf, 0, sizeof( buf ) );
-            memset( tmp, 0, sizeof( tmp ) );
-            mbedtls_ccm_setkey( &ccm, MBEDTLS_CIPHER_ID_AES, tmp, keysize );
-
-            TIME_AND_TSC( title,
-                    mbedtls_ccm_encrypt_and_tag( &ccm, BUFSIZE, tmp,
-                        12, NULL, 0, buf, buf, tmp, 16 ) );
-
-            mbedtls_ccm_free( &ccm );
+
+exit:
+    mbedtls_gcm_free(&gcm);
+#endif /* MBEDTLS_AES_C && MBEDTLS_GCM_C */
+
+    return ret;
+}
+
+MBED_NOINLINE static int benchmark_aes_ccm() {
+    int ret = 0;
+#if defined(MBEDTLS_AES_C) && defined(MBEDTLS_CCM_C)
+    int keysize;
+    mbedtls_ccm_context ccm;
+
+    mbedtls_ccm_init(&ccm);
+
+    for (keysize = 128; keysize <= 256; keysize += 64) {
+        ret = mbedtls_snprintf(title, sizeof(title), "AES-CCM-%d", keysize);
+        if (ret < 0 || static_cast<size_t>(ret) >= sizeof(title)) {
+            mbedtls_printf("Failed to compose title string using "
+                           "mbedtls_snprintf(): %d\n", ret);
+            goto exit;
         }
+
+        memset(buf, 0, sizeof(buf));
+        memset(tmp, 0, sizeof(tmp));
+
+        ret = mbedtls_ccm_setkey(&ccm, MBEDTLS_CIPHER_ID_AES, tmp, keysize);
+        if (ret != 0) {
+            mbedtls_printf("mbedtls_gcm_setkey() returned -0x%04X\n", -ret);
+            goto exit;
+        }
+
+        BENCHMARK_FUNC_CALL(title,
+                            mbedtls_ccm_encrypt_and_tag(&ccm, BUFSIZE, tmp, 12,
+                                                        NULL, 0, buf, buf, tmp,
+                                                        16));
     }
-#endif
-#if defined(MBEDTLS_CMAC_C)
-    if( todo->aes_cmac )
-    {
-        unsigned char output[16];
-        const mbedtls_cipher_info_t *cipher_info;
-        mbedtls_cipher_type_t cipher_type;
-        int keysize;
-
-        cipher_type = MBEDTLS_CIPHER_AES_128_ECB;
-        for( keysize = 128; keysize <= 256; keysize += 64 )
-        {
-            mbedtls_snprintf( title, sizeof( title ), "AES-CMAC-%d", keysize );
-
-            memset( buf, 0, sizeof( buf ) );
-            memset( tmp, 0, sizeof( tmp ) );
-
-            cipher_info = mbedtls_cipher_info_from_type( cipher_type );
-
-            TIME_AND_TSC( title,
-                          mbedtls_cipher_cmac( cipher_info, tmp, keysize,
-                                               buf, BUFSIZE, output ) );
-            cipher_type = (mbedtls_cipher_type_t)( cipher_type + 1 );
+
+exit:
+    mbedtls_ccm_free(&ccm);
+#endif /* MBEDTLS_AES_C && MBEDTLS_CCM_C */
+
+    return ret;
+}
+
+MBED_NOINLINE static int benchmark_aes_cmac() {
+    int ret = 0;
+#if defined(MBEDTLS_AES_C) && defined(MBEDTLS_CMAC_C)
```

Comment:
why not surround the whole function with `#if defined(MBEDTLS_AES_C) && defined(MBEDTLS_CMAC_C)`?